Add backup benchmarking under read stress

kreuzerkrieg opened the pull request:
Introduce the `test_backup_benchmark` test, which measures backup time under read stress conditions. This test performs multiple actions to consolidate all necessary data into a single table. Initially, it runs and measures the backup process, followed by the read stress test. Finally, it executes both processes asynchronously to observe how the performance of reading and backing up degrades.

More metrics could be added after https://github.com/scylladb/scylla-manager/issues/4123 is completed
Test should be re-ran after https://github.com/scylladb/scylla-manager/issues/4125 is completed

Argus results:
For [100GB run](https://argus.scylladb.com/tests/scylla-cluster-tests/83b4a96b-28bb-4501-8116-01f909250c03)
|  | backup time [s] |
| --- | --- |
| Backup during read stress | 00:18:48 |
| Backup | 00:10:18 |

|  | read time [s] |
| --- | --- |
| Read stress | 00:11:29 |
| Read stress during backup | 00:11:31 |

fixes: #8752

kreuzerkrieg commented:
> Introduce the `test_backup_benchmark` test, which measures backup time under read stress conditions. This test performs multiple actions to consolidate all necessary data into a single table. Initially, it runs and measures the backup process, followed by the read stress test. Finally, it executes both processes asynchronously to observe how the performance of reading and backing up degrades.
> 
> Argus results: For [100GB run](https://argus.scylladb.com/tests/scylla-cluster-tests/6a18ccd0-1d6a-4f00-a473-

I have to admit it looks super suspicious that the read did not degrade even slightly

mikliapko commented:
@kreuzerkrieg There was some [refactoring](https://github.com/scylladb/scylla-cluster-tests/pull/9222) done for Manager tests.
Please, consider rebasing to master and putting your new test to the right place or probably to a separate class for Backup benchmark.

kreuzerkrieg commented:
> @kreuzerkrieg There was some [refactoring](https://github.com/scylladb/scylla-cluster-tests/pull/9222) done for Manager tests. Please, consider rebasing to master and putting your new test to the right place or probably to a separate class for Backup benchmark.

rebased and moved the test to another class

mikliapko commented on mgmt_cli_test.py:
```diff
@@ -1550,3 +1551,88 @@ def test_restore_data_without_manager(self):
+
+            backup_start_time = time.time()
+            task = mgr_cluster.create_backup_task(location_list=self.locations, rate_limit_list=["0"])
+            backup_task_status = task.wait_for_uploading_stage(timeout=200000)
```
I think `200000` seconds timeout can be significantly decreased

kreuzerkrieg replied:
reduced

mikliapko commented on mgmt_cli_test.py:
```diff
@@ -1550,3 +1551,88 @@ def test_restore_data_without_manager(self):
+            backup_task_status = task.wait_for_uploading_stage(timeout=200000)
+            assert backup_task_status, "Backup has failed!"
+            backup_end_time = time.time()
+            InfoEvent(message=f'Backup without upload took: {backup_end_time - backup_start_time:.2f}s.').publish()
+            upload_start_time = time.time()
+            backup_status = task.wait_and_get_final_status(timeout=200000)
+            upload_end_time = time.time()
+            assert backup_status == TaskStatus.DONE, "Backup upload has failed!"
+            InfoEvent(
+                message=f'Backup upload upload took: {upload_end_time - upload_start_time}.').publish()
+            InfoEvent(
+                message=f'Backup total time is: {(backup_end_time - backup_start_time) + (upload_end_time - upload_start_time)}.').publish()
+            backup_report = {
+                "backup time": int(backup_end_time - backup_start_time),
+                "upload time": int(upload_end_time - upload_start_time),
+                "total": int((backup_end_time - backup_start_time) + (upload_end_time - upload_start_time)),
+            }
```
You may consider an option to get the whole backup process time as `task.duration` property.
It will minimize the number of calculations here

kreuzerkrieg replied:
Done

mikliapko commented on mgmt_cli_test.py:
```diff
@@ -1550,3 +1551,88 @@ def test_restore_data_without_manager(self):
+
+            for command in self.params.get('stress_read_cmd'):
+                stress_queue.append(self.run_stress_thread(command, round_robin=True, stop_test_on_failure=False))
+            read_start_time = time.time()
```
Btw, you may consider to use `ExecutionTimer` for these duration measurements:
https://github.com/scylladb/scylla-cluster-tests/blob/master/sdcm/utils/time_utils.py

kreuzerkrieg replied:
Done

mikliapko commented on mgmt_cli_test.py:
```diff
@@ -1550,3 +1551,88 @@ def test_restore_data_without_manager(self):
+            read_stress_report = {
+                "read stress time": int(read_end_time - read_start_time),
+            }
+            report_to_argus(read_stress_report, label)
```
I'm not sure `report_to_argus` should be a part of these `run_read_stress` and `create_backup` methods.
At least, method names are missing this part

kreuzerkrieg replied:
renamed

kreuzerkrieg commented on mgmt_cli_test.py:
```diff
@@ -1550,3 +1551,88 @@ def test_restore_data_without_manager(self):
+
+            backup_start_time = time.time()
+            task = mgr_cluster.create_backup_task(location_list=self.locations, rate_limit_list=["0"])
+            backup_task_status = task.wait_for_uploading_stage(timeout=200000)
+            assert backup_task_status, "Backup has failed!"
```
what do you mean? the `wait_for_uploading_stage` calls `task.wait_for_status` and it returns `bool` then I'll have to assert on `is_status_reached` or something

mikliapko replied:
I'm expecting `task.wait_for_status` raise `WaitForTimeoutError` exception if timeout reaches.
But whatever, the comment can be ignored, not significant at all.

Btw, your code might be a bit shortened:
```
assert task.wait_for_uploading_stage(timeout=200000), "Backup has failed!"
```
From my point of view, it improves readability but it's up to you.

Michal-Leszczynski commented on mgmt_cli_test.py:
```diff
@@ -1550,3 +1551,88 @@ def test_restore_data_without_manager(self):
+            mgr_cluster = self.ensure_and_get_cluster(
+                mgmt.get_scylla_manager_tool(manager_node=self.monitors.nodes[0]))
+
+            backup_start_time = time.time()
+            task = mgr_cluster.create_backup_task(location_list=self.locations, rate_limit_list=["0"])
+            backup_task_status = task.wait_for_uploading_stage(timeout=200000)
+            assert backup_task_status, "Backup has failed!"
+            backup_end_time = time.time()
```
Not really:/ The `sctool progress` allows you to check task status, which also includes task stage. You can check the current stage and see if SM if before, during or after snapshot creation, but it won't give you direct timestamp of snapshot creation.

```go
const (
	BackupStageInit         string = "INIT"
	BackupStageSnapshot     string = "SNAPSHOT"
	BackupStageAwaitSchema  string = "AWAIT_SCHEMA"
	BackupStageIndex        string = "INDEX"
	BackupStageManifest     string = "MANIFEST"
	BackupStageSchema       string = "SCHEMA"
	BackupStageDeduplicate  string = "DEDUPLICATE"
	BackupStageUpload       string = "UPLOAD"
	BackupStageMoveManifest string = "MOVE_MANIFEST"
	BackupStageMigrate      string = "MIGRATE"
	BackupStagePurge        string = "PURGE"
	BackupStageDone         string = "DONE"
)
```

regevran commented:
The units in the issue should be ~[s]~ --> [h:m:s]

kreuzerkrieg commented on sdcm/argus_results.py:
```diff
@@ -106,6 +106,9 @@ class Meta:
             ColumnMetadata(name="repair time", unit="s", type=ResultType.DURATION, higher_is_better=False),
+            ColumnMetadata(name="backup time", unit="s", type=ResultType.DURATION, higher_is_better=False),
+            ColumnMetadata(name="upload time", unit="s", type=ResultType.DURATION, higher_is_better=False),
+            ColumnMetadata(name="read stress time", unit="s", type=ResultType.DURATION, higher_is_better=False),
```
Added two new result tables

kreuzerkrieg commented on mgmt_cli_test.py:
```diff
@@ -1550,3 +1551,88 @@ def test_restore_data_without_manager(self):
+            read_stress_report = {
+                "read stress time": int(read_end_time - read_start_time),
+            }
+            report_to_argus(read_stress_report, label)
```
Done

kreuzerkrieg commented:
> The units in the issue should be ~[s]~ --> [h:m:s]

It is the way it is presented, I think, I do send seconds to it and it presents it as a time

kreuzerkrieg commented:
> backup time [s]	upload time [s]	total [s]
> Backup times	00:00:12	00:09:45	00:09:11
> Backup during read stress	00:00:12	00:15:07	00:14:43
> From the code I see that `backup time` is total backup time without upload, right? It would be nice to indicate it with a more descriptive column name.

Should I name it "snapshot time"?

> Also, why `total` is smaller than `upload time`?

Good question, the total time is taken from backup `task.duration` the rest is measured in place and it doesnt add up. Ideas?

Michal-Leszczynski commented:
> Should I name it "snapshot time"?

It's not only about snapshot (e.g. we also fetch schema, create manifests, etc...).
In general I'm not sure why do we need to report the time up to the upload stage? Is it useful?
But I see that the upload time is also not only about upload (e.g. purging backup location from unnecessary files).
I will leave some comments in changed files in a bit.

I created and [issue](https://github.com/scylladb/scylla-manager/issues/4123) about displaying backup upload bandwidth/duration - in the future it could make things easier for such benchmarks.

Michal-Leszczynski commented on mgmt_cli_test.py:
```diff
@@ -1550,3 +1555,101 @@ def test_restore_data_without_manager(self):
+        submit_results_to_argus(self.test_config.argus_client(), table)
+
+    def create_backup_and_report(self, mgr_cluster, label: str):
+        task = mgr_cluster.create_backup_task(location_list=self.locations, rate_limit_list=["0"])
```
In terms of running the backup as fast as possible (and with as much cluster pressure as possible), SM 3.4 introduced `--transfers` flag which controls the amount of parallel file transfers (2 by default).

When downloading files during restore, we got the best results when setting it to `2*node_shard_cnt`.
Not sure if the same value would be the best for upload, but increasing it from 2 could definitely help.

regevran replied:
We need a way to accurately and _repeatedly_ measure the same thing.
If "cluster pressure" means lots of reads from disk during backup - then yes, this is what we need.

kreuzerkrieg replied:
tried to increase it to 16 and then to 32 and I dont see any difference. Moreover, the second run of backup timeouts since for some unknown reason the backup proves become significantly slower even when the concurrent read is over long time ago

kreuzerkrieg replied:
I will keep track of it, once [the issue](https://github.com/scylladb/scylla-manager/issues/4125) is resolved I will try to rerun it with different `--transfers` settings

Michal-Leszczynski commented on configurations/manager/100GB_dataset.yaml:
```diff
@@ -0,0 +1,14 @@
+                     "cassandra-stress write cl=ALL n=26214400 -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native  -rate threads=500 -col 'size=FIXED(1024) n=FIXED(1)' -pop seq=52428801..78643200",
+                     "cassandra-stress write cl=ALL n=26214400 -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native  -rate threads=500 -col 'size=FIXED(1024) n=FIXED(1)' -pop seq=78643201..104857600" ]
+
+stress_read_cmd: [ "cassandra-stress read cl=ALL n=26214400 -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native  -rate threads=500 -col 'size=FIXED(1024) n=FIXED(1)' -pop seq=1..26214400",
+                   "cassandra-stress read cl=ALL n=26214400 -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native  -rate threads=500 -col 'size=FIXED(1024) n=FIXED(1)' -pop seq=26214401..52428800",
+                   "cassandra-stress read cl=ALL n=26214400 -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native  -rate threads=500 -col 'size=FIXED(1024) n=FIXED(1)' -pop seq=52428801..78643200",
+                   "cassandra-stress read cl=ALL n=26214400 -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native  -rate threads=500 -col 'size=FIXED(1024) n=FIXED(1)' -pop seq=78643201..104857600" ]
```
Maybe we would notice read degradation if we increase the read pressure?
I'm also thinking about changing it to mixed reads-writes, but then I'm not sure about how disabled auto compaction impacts this part.

regevran replied:
We might want to add writes later on, but for the base measures reads would be enough.
We should see degradation in times for both the upload and the queries.
I.e - backing up 100GB and reading 100GB at the same time should both take longer compared to just uploading and to just reading respectively.

kreuzerkrieg replied:
> Maybe we would notice read degradation if we increase the read pressure? I'm also thinking about changing it to mixed reads-writes, but then I'm not sure about how disabled auto compaction impacts this part.

quite the opposite, I'm running now with 50 threads just to get to reasonable latencies, 500 thread is such an overkill that the latency is skyrocketing and the CPU is 100% utilized.
with 50 threads I'm getting ~1.5ms latency and I expect it rise during the concurrent backup is running

kreuzerkrieg commented:
> > Should I name it "snapshot time"?
> 
> It's not only about snapshot (e.g. we also fetch schema, create manifests, etc...). In general I'm not sure why do we need to report the time up to the upload stage? Is it useful? But I see that the upload time is also not only about upload (e.g. purging backup location from unnecessary files). I will leave some comments in changed files in a bit.
> 
> I created and [issue](https://github.com/scylladb/scylla-manager/issues/4123) about displaying backup upload bandwidth/duration - in the future it could make things easier for such benchmarks.

Now when I see that it is negligible slice of the whole process I guess it is worth just dropping it

kreuzerkrieg commented on mgmt_cli_test.py:
```diff
@@ -1550,3 +1555,101 @@ def test_restore_data_without_manager(self):
+        task = mgr_cluster.create_backup_task(location_list=self.locations, rate_limit_list=["0"])
+
+        with ExecutionTimer() as backup_timer:
+            backup_task_status = task.wait_for_uploading_stage(timeout=600)
+            assert backup_task_status, "Backup has failed!"
+        InfoEvent(message=f'Backup without upload took: {backup_timer.duration}s.').publish()
+
+        with ExecutionTimer() as backup_upload_timer:
+            backup_status = task.wait_and_get_final_status(timeout=72000)
```
same as https://github.com/scylladb/scylla-cluster-tests/pull/9307#issuecomment-2498280918

mikliapko commented on sdcm/argus_results.py:
```diff
@@ -117,6 +117,24 @@ class Meta:
         }
 
 
+class BackupResult(GenericResultTable):
+    class Meta:
+        name = "Backup benchmark"
+        description = "Backup benchmark"
+        Columns = [
+            ColumnMetadata(name="backup time", unit="s", type=ResultType.DURATION, higher_is_better=False),
+        ]
+
+
+class ReadResult(GenericResultTable):
+    class Meta:
+        name = "Read timing"
+        description = "Read timing"
+        Columns = [
+            ColumnMetadata(name="read time", unit="s", type=ResultType.DURATION, higher_is_better=False),
+        ]
```
I'd suggest to show connection to Manager in class names.
What do you think about renaming to `ManagerBackupBenchmarkResult` and `ManagerBackupBenchmarkReadResult`?
Yes, it's more complex but pretty straightforward for everyone.

kreuzerkrieg replied:
I dont mind having whatever name for the class, but why it should be connected to the Manages? Is it possible that someone would use this table to report read timing to argus not in a context of manager?

mikliapko replied:
Good question btw.
@soyacz What approach do you prefer?

soyacz replied:
Let's add `ManagerBackup` prefix as currently it sounds too generic. 
Unless you decide to refactor and put Manager related stuff out of this file to separate one somewhere around manager code - then prefixes are redundant.

Btw. from my uninvolved perspective I don't know even what is `read` - does it refer to backup or reading from db? - descriptions could be more descriptive so anyone could understand it.

kreuzerkrieg replied:
Done

mikliapko commented on mgmt_cli_test.py:
```diff
@@ -1550,3 +1555,93 @@ def test_restore_data_without_manager(self):
         self.test_restore_from_precreated_backup(snapshot_name, restore_outside_manager=True)
+
+
+class ReportType(Enum):
```
And the same here about class name, please

kreuzerkrieg replied:
Done

kreuzerkrieg commented:
@scylladb/qa-maintainers could you push it forward please?

kreuzerkrieg commented:
> I briefly looked at 100gb dataset - for me ingestion of this size should take bit more than 7m - did you verify correctness of this element?

This test uses existing ingestion machinery. I assume that ingestion works.
In any case, looks like it takes mpore or less 7 or 8 minutes
![image](https://github.com/user-attachments/assets/98c56d3c-50e7-4982-9a7a-124a1b34a384)


> also there's no jenkinsfile related this test, how it is going to be used?

As any other SM test, just drop the right test name in the existing jenkins pipeline

![image](https://github.com/user-attachments/assets/0d7085e3-e5ba-4d68-b60d-79ab6ec9eeeb)
